### PR TITLE
Introduce CircuitHandler to handle circuit lifetime events

### PR DESCRIPTION
**Pull request by pranavkm**
Partial fix to https://github.com/aspnet/AspNetCore/issues/6353

### CODE REVIEW

File: src/Components/Server/src/Builder/RazorComponentsApplicationBuilderExtensions.cs [hunk 1]
```diff
@@ -50,7 +50,7 @@ public static class RazorComponentsApplicationBuilderExtensions
             // add SignalR and BlazorHub automatically.
             if (options.UseSignalRWithBlazorHub)
             {
-                builder.UseSignalR(route => route.MapHub<BlazorHub>(BlazorHub.DefaultPath));
+                builder.UseSignalR(route => route.MapHub<ComponentsHub>(ComponentsHub.DefaultPath));
```
Comment (pranavkm) [inline src/Components/Server/src/Builder/RazorComponentsApplicationBuilderExtensions.cs]: Cleaning up some things as I encounter them.

File: src/Components/Server/src/ComponentsServerOptions.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+{
+    public class ComponentsServerOptions
+    {
+        private readonly Dictionary<PathString, Type> _circuitHandlers = new Dictionary<PathString, Type>();
+
+        public IReadOnlyDictionary<PathString, Type> CircuitHandlers => _circuitHandlers;
+
+        public Type DefaultCircuitHandler { get; private set; }
+
+        public void AddCircuitHandler<THandler>(PathString path) where THandler : CircuitHandler
```
Comment (pranavkm) [inline src/Components/Server/src/ComponentsServerOptions.cs]: This thing is fairly obtuse, but it sort of follows the same pattern as `StartupActions`. I couldn't figure out a way to associate each hub instance with a separate configuration. 

For features like tombstoning, it's likely that users are going to need other DI services in their handlers. I wanted to see if we could force a design that required doing a DI. But the API does feel very user hostile.

File: src/Components/Server/src/Circuits/CircuitHost.cs [hunk 3]
```diff
@@ -126,9 +134,15 @@ public async void BeginInvokeDotNetFromJS(string callId, string assemblyName, st
             }
         }
 
-        public void Dispose()
+        public async ValueTask DisposeAsync()
```
Comment (pranavkm) [inline src/Components/Server/src/Circuits/CircuitHost.cs]: Weird to use DisposeAsync for this? Should we have an explicit event for disconnects?

File: src/Components/Server/src/Circuits/CircuitHandler.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Components.Server.Circuits
+{
+    /// <summary>
+    /// A <see cref="CircuitHandler"/> allows running code during specific lifetime events of a <see cref="Circuit"/>.
+    /// <list type="bullet">
+    /// <item>
+    /// <see cref="OnCircuitOpenedAsync(Circuit, CancellationToken)"/> is invoked after an initial connection to the client
```
Comment (rynowak) [inline src/Components/Server/src/Circuits/CircuitHandler.cs]: s/connection/circuit

File: src/Components/Server/src/Circuits/CircuitHandler.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    /// has been established.
+    /// </item>
+    /// <item>
+    /// <see cref="OnConnectionUpAsync(Circuit, CancellationToken)(Circuit, CancellationToken)"/> is invoked immediately after the completion of
+    /// <see cref="OnCircuitOpenedAsync(Circuit, CancellationToken)"/>. In addition, the method is invoked each time a connection is re-established
+    /// with a client after it's been dropped. <see cref="OnConnectionDownAsync(Circuit, CancellationToken)"/> is invoked each time a connection
+    /// is dropped.
+    /// </item>
+    /// <item>
+    /// <see cref="OnCircuitClosedAsync(Circuit, CancellationToken)"/> is invoked prior to the server evicting the connection to the client.
```
Comment (rynowak) [inline src/Components/Server/src/Circuits/CircuitHandler.cs]: s/connection/circuit

File: src/Components/Server/src/Circuits/CircuitHandler.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    {
+        internal static readonly CircuitHandler NullHandler = new NullCircuitHandler();
+
+        /// <summary>
+        /// Invoked when a new circuit was established.
+        /// </summary>
+        /// <param name="circuit">The <see cref="Circuit"/>.</param>
+        /// <param name="cts">The <see cref="CancellationToken"/>.</param>
+        /// <returns><see cref="Task"/> that represents the asynchronous execution operation.</returns>
+        public virtual Task OnCircuitOpenedAsync(Circuit circuit, CancellationToken cts) => Task.CompletedTask;
```
Comment (rynowak) [inline src/Components/Server/src/Circuits/CircuitHandler.cs]: `cts` -> `cancellationToken = default`

File: src/Components/Server/src/Circuits/CircuitHost.cs [hunk 7]
```diff
@@ -37,25 +40,24 @@ public static void SetCurrentCircuitHost(CircuitHost circuitHost)
...
             IServiceScope scope,
             IClientProxy client,
             RendererRegistry rendererRegistry,
             RemoteRenderer renderer,
             Action<IComponentsApplicationBuilder> configure,
             IJSRuntime jsRuntime,
-            CircuitSynchronizationContext synchronizationContext)
+            CircuitSynchronizationContext synchronizationContext,
+            Circuit circuit,
+            CircuitHandler circuitHandler)
```
Comment (rynowak) [inline src/Components/Server/src/Circuits/CircuitHost.cs]: Is there only a single handler?

Comment (rynowak) [reply src/Components/Server/src/Circuits/CircuitHost.cs]: Do you think that this is an important disinction?

File: src/Components/Server/src/Circuits/CircuitHost.cs [hunk 8]
```diff
@@ -65,7 +67,8 @@ public static void SetCurrentCircuitHost(CircuitHost circuitHost)
 
             Services = scope.ServiceProvider;
 
-            Circuit = new Circuit(this);
```
Comment (rynowak) [inline src/Components/Server/src/Circuits/CircuitHost.cs]: Does this super need to change? It's intentional that the circuit is **dumb** - I constructed it here to ensure that it stays that way.

File: src/Components/Server/src/Circuits/CircuitHost.cs [hunk 9]
```diff
@@ -102,6 +107,9 @@ public async Task InitializeAsync()
                     var entry = builder.Entries[i];
                     Renderer.AddComponent(entry.componentType, entry.domElementSelector);
                 }
+
+                await CircuitHandler.OnCircuitOpenedAsync(Circuit, cancellationToken);
+                await CircuitHandler.OnConnectionUpAsync(Circuit, cancellationToken);
```
Comment (rynowak) [inline src/Components/Server/src/Circuits/CircuitHost.cs]: Dope

File: src/Components/Server/src/Circuits/CircuitHost.cs (hunk 3 above)
Comment (rynowak) [reply src/Components/Server/src/Circuits/CircuitHost.cs]: This seems cool.

File: src/Components/Server/src/Circuits/DefaultCircuitFactory.cs [hunk 10]
```diff
@@ -14,11 +14,11 @@ namespace Microsoft.AspNetCore.Components.Server.Circuits
     internal class DefaultCircuitFactory : CircuitFactory
     {
         private readonly IServiceScopeFactory _scopeFactory;
-        private readonly DefaultCircuitFactoryOptions _options;
+        private readonly ComponentsServerOptions _options;
```
Comment (rynowak) [inline src/Components/Server/src/Circuits/DefaultCircuitFactory.cs]: eh this name is uhhhhhhhhhhhhhhh

File: src/Components/Server/src/Circuits/DefaultCircuitFactoryOptions.cs [hunk 11]
```diff
@@ -1,22 +0,0 @@
...
-// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
-
-using System;
-using System.Collections.Generic;
-using Microsoft.AspNetCore.Components.Builder;
-using Microsoft.AspNetCore.Http;
-
-namespace Microsoft.AspNetCore.Components.Server.Circuits
-{
-    internal class DefaultCircuitFactoryOptions
```
Comment (rynowak) [inline src/Components/Server/src/Circuits/DefaultCircuitFactoryOptions.cs]: Can we change this back?

File: src/Components/Server/src/ComponentsHub.cs [hunk 12]
```diff
@@ -25,18 +25,27 @@ public sealed class BlazorHub : Hub
...
         /// </summary>
-        public BlazorHub(
-            ILogger<BlazorHub> logger,
-            IServiceProvider services)
+        public ComponentsHub(
+            IServiceProvider services,
+            ILogger<ComponentsHub> logger)
+            : this(
+                  services.GetRequiredService<CircuitFactory>(),
+                  logger)
```
Comment (rynowak) [inline src/Components/Server/src/ComponentsHub.cs]: This formatting is stupid

File: src/Components/Server/src/ComponentsHub.cs [hunk 13]
```diff
@@ -66,7 +74,8 @@ public async Task StartCircuit(string uriAbsolute, string baseUriAbsolute)
 
             // If initialization fails, this will throw. The caller will explode if they
             // try to call into any interop API.
-            await circuitHost.InitializeAsync();
+            await circuitHost.InitializeAsync(Context.ConnectionAborted);
+
```
Comment (rynowak) [inline src/Components/Server/src/ComponentsHub.cs]: This is really nice and clean. I like how this didn't get way more complicated

File: src/Components/Server/src/ComponentsServerOptions.cs [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Components.Builder;
+using Microsoft.AspNetCore.Components.Server.Circuits;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Components.Server
+{
+    public class ComponentsServerOptions
```
Comment (rynowak) [inline src/Components/Server/src/ComponentsServerOptions.cs]: I'd love to find a way to get rid of the plural in this name.

File: src/Components/Server/src/ComponentsServerOptions.cs [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+        public void SetDefaultHandler<THandler>() where THandler : CircuitHandler
+        {
+            DefaultCircuitHandler = typeof(THandler);
+        }
+
+        // During the DI configuration phase, we use Configure<DefaultCircuitFactoryOptions>(...)
+        // callbacks to build up this dictionary mapping paths to startup actions
+        internal Dictionary<PathString, Action<IComponentsApplicationBuilder>> StartupActions { get; }
+            = new Dictionary<PathString, Action<IComponentsApplicationBuilder>>();
+    }
```
Comment (rynowak) [inline src/Components/Server/src/ComponentsServerOptions.cs]: Do we need this at all? What if we used DI instead? I'm not sure you'd want these things to behave like MVC filters. I think the norm is to have a single entry point or treat every entry point the same. I don't think we need filtering by path.

**Review by rynowak — COMMENTED** [review]
This seems super on the right track.

File: src/Components/Server/src/Circuits/CircuitHandler.cs [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+        /// and once each for each reconnect during the lifetime of a circuit.
+        /// </para>
+        /// </summary>
+        /// <param name="circuit">The <see cref="Circuit"/>.</param>
+        /// <param name="cts">The <see cref="CancellationToken"/>.</param>
+        /// <returns><see cref="Task"/> that represents the asynchronous execution operation.</returns>
+        public virtual Task OnConnectionUpAsync(Circuit circuit, CancellationToken cts) => Task.CompletedTask;
+
+        /// <summary>
+        /// Invoked a connection to the client using was dropped.
```
Comment (javiercn) [inline src/Components/Server/src/Circuits/CircuitHandler.cs]: Grammar?

File: src/Components/Server/src/ComponentsHub.cs (hunk 12 above)
Comment (javiercn) [reply src/Components/Server/src/ComponentsHub.cs]: Agree

File: src/Components/test/testassets/ComponentsApp.App/ComponentsApp.App.csproj [hunk 17]
```diff
@@ -1,8 +1,10 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk.Razor">
 
+  <!-- Reference targets and props from Components.Build package -->
+  <Import Project="..\..\..\Build\src\ReferenceFromSource.props" />
+
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
-    <ReferenceBlazorBuildLocally>true</ReferenceBlazorBuildLocally>
```
Comment (natemcmaster) [inline src/Components/test/testassets/ComponentsApp.App/ComponentsApp.App.csproj]: Why did you remove `ReferenceBlazorBuildLocally` and replace it with a direct import?

**Comment by pranavkm** [comment]
🆙 📅

File: src/Components/test/testassets/ComponentsApp.App/ComponentsApp.App.csproj (hunk 17 above)
Comment (natemcmaster) [reply src/Components/test/testassets/ComponentsApp.App/ComponentsApp.App.csproj]: Ah. I didn't notice there was a difference.

File: src/Components/Server/src/Circuits/CircuitHandler.cs (hunk 6 above)
Comment (rynowak) [reply src/Components/Server/src/Circuits/CircuitHandler.cs]: Please fix the parameter name.

**Review by rynowak — DISMISSED** [review]
https://github.com/aspnet/AspNetCore/pull/6971/files#r251153473

**Comment by Tratcher** [comment]
Weird, I don't see why I was triggered as a code owner.

File: eng/Versions.props [hunk 18]
```diff
@@ -145,7 +145,7 @@
     <SystemIdentityModelTokensJwtPackageVersion>5.3.0</SystemIdentityModelTokensJwtPackageVersion>
     <WindowsAzureStoragePackageVersion>8.1.4</WindowsAzureStoragePackageVersion>
     <!-- Dependencies for Blazor. -->
-    <MicrosoftAspNetCoreBlazorMonoPackageVersion>0.8.0-preview1-20181126.1</MicrosoftAspNetCoreBlazorMonoPackageVersion>
+    <MicrosoftAspNetCoreBlazorMonoPackageVersion>0.8.0-preview-20190125.1</MicrosoftAspNetCoreBlazorMonoPackageVersion>
```
Comment (rynowak) [inline eng/Versions.props]: I see that you've got a game of deadly cat and mouse going here. Just make sure you don't break this when you merge 😆

File: src/Components/Server/src/Circuits/CircuitHost.cs [hunk 19]
```diff
@@ -81,30 +87,40 @@ public static void SetCurrentCircuitHost(CircuitHost circuitHost)
...
         public RendererRegistry RendererRegistry { get; }
 
-        public IServiceScope Scope { get; }
-
         public IServiceProvider Services { get; }
 
         public CircuitSynchronizationContext SynchronizationContext { get; }
 
-        public async Task InitializeAsync()
+        public CancellationToken ConnectionAborted { get; }
```
Comment (rynowak) [inline src/Components/Server/src/Circuits/CircuitHost.cs]: Unused?

File: src/Components/Server/src/Circuits/DefaultCircuitFactoryOptions.cs [hunk 20]
```diff
@@ -6,17 +6,13 @@
 using Microsoft.AspNetCore.Components.Builder;
 using Microsoft.AspNetCore.Http;
 
-namespace Microsoft.AspNetCore.Components.Server.Circuits
+namespace Microsoft.AspNetCore.Components.Server
 {
-    internal class DefaultCircuitFactoryOptions
+    public class DefaultCircuitFactoryOptions
```
Comment (rynowak) [inline src/Components/Server/src/Circuits/DefaultCircuitFactoryOptions.cs]: Why is this public?

**Comment by pranavkm** [comment]
🆙 📅

File: src/Components/Server/src/Circuits/CircuitHandler.cs [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+        /// <value>
+        /// Defaults to 0.
+        /// </value>
+        public virtual int Order => 0;
+
+        /// <summary>
+        /// Invoked when a new circuit was established.
+        /// </summary>
+        /// <param name="circuit">The <see cref="Circuit"/>.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/>.</param>
```
Comment (SteveSandersonMS) [inline src/Components/Server/src/Circuits/CircuitHandler.cs]: Might be nice to specify what this particular `CancellationToken` represents. What operation does it represent the cancellation of? Also in the other methods in this class.

**Review by SteveSandersonMS — APPROVED** [review]
This looks superb to me.

File: src/Components/Server/src/Circuits/Circuit.cs [hunk 22]
```diff
@@ -1,35 +1,23 @@
...
-using Microsoft.JSInterop;
-
 namespace Microsoft.AspNetCore.Components.Server.Circuits
 {
     /// <summary>
-    /// Represents an active connection between a Blazor server and a client.
+    /// Represents a link between a ASP.NET Core Component on the server and a client.
     /// </summary>
-    public class Circuit
+    public sealed class Circuit
```
Comment (javiercn) [inline src/Components/Server/src/Circuits/Circuit.cs]: What's the purpose of this class? It only has a single ID and a _circuitHost that is not accessible outside this class. Can't the Id be moved to the CircuitHost class and this class be removed?

File: src/Components/Server/src/Circuits/CircuitHost.cs [hunk 23]
```diff
@@ -37,25 +42,23 @@ public static void SetCurrentCircuitHost(CircuitHost circuitHost)
...
             RendererRegistry rendererRegistry,
             RemoteRenderer renderer,
             Action<IComponentsApplicationBuilder> configure,
             IJSRuntime jsRuntime,
-            CircuitSynchronizationContext synchronizationContext)
+            CircuitSynchronizationContext synchronizationContext,
+            CircuitHandler[] circuitHandlers)
         {
-            Scope = scope ?? throw new ArgumentNullException(nameof(scope));
+            _scope = scope ?? throw new ArgumentNullException(nameof(scope));
```
Comment (javiercn) [inline src/Components/Server/src/Circuits/CircuitHost.cs]: I think at some point we should do cleanup of our DI usage across circuits and Components. We shouldn't be passing DI abstractions around unless the specific class is a composition root.

**Review by javiercn — APPROVED** [review]

File: src/Components/Server/src/Circuits/Circuit.cs (hunk 22 above)
Comment (pranavkm) [reply src/Components/Server/src/Circuits/Circuit.cs]: I expect `Circuit` to be the user facing type. `CircuitHost` is an implementation detail.

File: src/Components/Server/src/Circuits/CircuitHost.cs [hunk 24]
```diff
@@ -66,11 +69,14 @@ public static void SetCurrentCircuitHost(CircuitHost circuitHost)
             Services = scope.ServiceProvider;
 
             Circuit = new Circuit(this);
+            _circuitHandlers = circuitHandlers;
 
             Renderer.UnhandledException += Renderer_UnhandledException;
             SynchronizationContext.UnhandledException += SynchronizationContext_UnhandledException;
         }
 
+        public string CircuitId { get; } = Guid.NewGuid().ToString();
```
[the remark this answers is not in the document]
Comment (pranavkm) [reply src/Components/Server/src/Circuits/CircuitHost.cs]: Tracking via https://github.com/aspnet/AspNetCore/issues/7243

File: src/Components/Server/src/Circuits/CircuitHost.cs (hunk 23 above)
Comment (rynowak) [reply src/Components/Server/src/Circuits/CircuitHost.cs]: The circuit host is the **owner** of the scope which is why it's here.
